File: 백준/Silver/1991. 트리 순회/트리 순회.py
import sys
input=sys.stdin.readline
from collections import defaultdict

n = int(input())
lc = defaultdict(str)
rc = defaultdict(str)
for _ in range(n):
    p,l,r = input().rstrip().split()
    if l != '.':
        lc[p] = l
    if r != '.':
        rc[p] = r

def preorder(s):
    print(s,end="")
    if lc[s]:
        preorder(lc[s])
    if rc[s]:
        preorder(rc[s])

def inorder(s):
    if lc[s]:
        inorder(lc[s])
    print(s,end="")
    if rc[s]:
        inorder(rc[s])
    
def postorder(s):
    if lc[s]:
        postorder(lc[s])
    if rc[s]:
        postorder(rc[s])
    print(s,end="")

preorder('A')
print()
inorder('A')
print()
postorder('A')

File: 백준/Silver/1991. 트리 순회/트리 순회.java
import java.io.*;
import java.util.*;

public class Main {
	static int[][] tree;
	static StringBuilder sb;

	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		tree = new int[26][2];
		StringTokenizer st;
		sb = new StringBuilder();
		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			int a = st.nextToken().charAt(0) - 'A';
			char b = st.nextToken().charAt(0);
			char c = st.nextToken().charAt(0);
			if (b != '.') {
				tree[a][0] = b - 'A';
			}
			if (c != '.') {
				tree[a][1] = c - 'A';
			}
		}
		preorder(0);
		sb.append("\n");
		inorder(0);
		sb.append("\n");
		postorder(0);
		System.out.print(sb);
	}

	static void preorder(int x) {
		sb.append((char) (x + 'A'));
		if (tree[x][0] != 0)
			preorder(tree[x][0]);
		if (tree[x][1] != 0)
			preorder(tree[x][1]);
	}

	static void inorder(int x) {
		if (tree[x][0] != 0)
			inorder(tree[x][0]);
		sb.append((char) (x + 'A'));
		if (tree[x][1] != 0)
			inorder(tree[x][1]);
	}

	static void postorder(int x) {
		if (tree[x][0] != 0)
			postorder(tree[x][0]);
		if (tree[x][1] != 0)
			postorder(tree[x][1]);
		sb.append((char) (x + 'A'));
	}
}